R/abs_api.R

```diff
@@ -0,0 +1,220 @@
+  desc <- purrr::map_chr(desc, ~ifelse(is.null(.), "", .))
+  out$desc <- desc
+
+  out[c("id", "name", "desc", "version")]
+}
+
+
+#' @export
+#' @rdname abs_api
+abs_data <- function(id, datakey = "all", start_period = NULL, end_period = NULL, version = NULL) {
+  # Build query
+  dataflow <- paste0("ABS,", id)
+  if (!is.null(version)) dataflow <- paste(dataflow, version, sep=",")
+  q <- c("startPeriod" = start_period, "endPeriod" = end_period)
+  url <- abs_api_url(c("data", dataflow, datakey), q)
+
+  # Query API for labels and data
+  codelist <- abs_datastructure(id)
+  codelist <- codelist[!is.na(codelist$code), ]
+
+  as_csv <- httr::accept("application/vnd.sdmx.data+csv")
+  r <- httr::GET(url, as_csv)
+  httr::stop_for_status(r)
+  r <- resp_as_df(r)
+
+  # Label values
+  r <- purrr::imap_dfc(r, function(x, var_name) {
+    codes <- codelist[codelist$var == var_name, ]
+    if (nrow(codes)==0) return(x)
+    labs <- codes$code
+
+    # Match class avoiding data loss
+    if (can_numeric(x) & can_numeric(labs)) {
+      x <- as.numeric(x)
+      labs <- as.numeric(labs)
+    } else {
+      x <- as.character(x)
+      labs <- as.character(labs)
+    }
+
+    names(labs) <- codes$label
+    labelled::labelled(x, labs, unique(codes$desc))
+  })
+
+  r[setdiff(names(r), "DATAFLOW")]
+}
+
+#' @export
+#' @rdname abs_api
+abs_datastructure <- function(id) {
+  r <- httr::GET(abs_api_url(c("datastructure", "ABS", id, "?references=codelist")), httr::accept_xml())
+  httr::stop_for_status(r)
+  r <- httr::content(r)
+
+  r <-  xml2::as_list(r)
+
+  codelists <- r$Structure$Structures$Codelists
+  names(codelists) <- purrr::map_chr(r$Structure$Structures$Codelists, ~attr(., "id"))
+  components <-  r$Structure$Structures$DataStructures$DataStructure$DataStructureComponents
+
+  codelists <- purrr::imap_dfr(codelists, function(codelist, nm) {
+    desc <- codelist$Name[[1]]
+    purrr::map_dfr(codelist, ~c(local_id = nm, desc = desc, code = attr(., "id"), label = .$Name))
+  })
+  codelists <- codelists[!is.na(codelists$code), ]
+
+  components <- purrr::imap_dfr(components, function(x, nm) {
+    nm <- tolower(gsub("List", "", nm))
+    purrr::map_dfr(x, ~c(
+      role = nm,
+      var = attr(., "id"),
+      local_id = attr(.$LocalRepresentation$Enumeration$Ref, "id"),
+      position = attr(., "position")
+    ))
+  })
+
+  out <- merge(components, codelists, by = "local_id", all.x = TRUE, all.y = TRUE)
+  out <- out[order(out$code), ]
+  out <- out[order(out$position), setdiff(names(out), "local_id")]
+  rownames(out) <- NULL
+
+  out
+}
+
+
+# Internal ---------------------------------------------------------------------
+
+#' Convert an httr response containing a csv into a df
+#'
+#' @param r An httr response
+#'
+#' @return A data.frame
+#' @noRd
+#' @keywords internal
+#'
+resp_as_df <- function(r) {
+  fp <- tempfile()
+  writeLines(rawToChar(httr::content(r)), fp)
+  r <- read.csv(fp)
```

Comment:
Consider the `text` argument to `read.csv` instead of the round trip to disk. `read.csv(text = rawToChar(..`

Reply:
Oh great call! Thanks for that